core/src/main/java/com/linecorp/armeria/internal/common/util/ObjectCollectingUtil.java

```diff
@@ -107,6 +107,12 @@
 
         @Override
         public void onSubscribe(Subscription s) {
+            future.handle((ignored, cause) -> {
+                if (cause != null) {
+                    s.cancel();
```

Comment:
How about adding a boolean flag `onErrorCalled`? so that we don't unnecessarily call `s.cancel()`?

Reply:
Added the flag as per the suggestion.
I had the flag as a volatile field because `onError` may get called in a different thread than the subscription's thread.

Please, correct me if I am wrong 🙏.

Reply:
We don't need the volatile flag because the thread that calls `future.completeExceptionally(...)` also invokes `future.handle(...)`. 😉

Reply:
Ah, callback is not executed in the subscription thread, thanks!!